### PR TITLE
Amended data structures

### PR DESCRIPTION
**Pull request by dguhl**
According to [PHPSTAN](https://github.com/phpstan/phpstan/issues/3470), some methods were missing in the Map class and others. I amended my contribution.

PS: I closed my original PR #832 in that matter because it continued the old branch on my side with a cluttered commit history. I opened a new branch to keep the contribution clean.

### CODE REVIEW

File: ds/ds.php [hunk 1]
```diff
@@ -842,6 +842,16 @@ public function jsonSerialize()
 
     class Deque implements Sequence
     {
+        /**
+         * Creates a new instance, using either a traversable object or an array for the initial values.
+         * @param mixed ...$values A traversable object or an array to use for the initial values.
+         *
+         * @link https://www.php.net/manual/en/ds-map.construct.php
```
Comment (isfedorov) [inline ds/ds.php:849]: Seems this link is for Map class but not Deque. And correct link for Deque class is https://www.php.net/manual/en/ds-deque.construct.php

File: ds/ds.php [hunk 2]
```diff
@@ -1242,6 +1252,50 @@ public function jsonSerialize()
 
     class Map implements Collection
     {
+        /**
+         * Creates a new instance, using either a traversable object or an array for the initial values.
+         * @param mixed ...$values A traversable object or an array to use for the initial values.
```
Comment (isfedorov) [inline ds/ds.php:1257]: Please add correct doc link for constructor of Map class

File: ds/ds.php [hunk 3]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+         * @link https://www.php.net/manual/en/ds-map.diff.php
+         */
+        public function diff(Map $map): Map
+        {
+        }
+
+        /**
+         * Creates a new map using a callable to determine which pairs to include
+         *
+         * @param callable $callback Optional callable which returns TRUE if the pair should be included, FALSE
```
Comment (isfedorov) [inline ds/ds.php:1349]: please replace Optional to [optional] to make param interpreted as optional

File: ds/ds.php [hunk 4]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+         * otherwise. If a callback is not provided, only values which are TRUE (see converting to boolean) will be included.
+         *
+         * @return Map
+         *
+         * @link https://www.php.net/manual/en/ds-map.filter.php
+         */
+        public function filter(callable $callback): Map
```
Comment (isfedorov) [inline ds/ds.php:1356]: Please add `null` as default value for callback to get rid of "Hierarchy check" inspection warning

File: ds/ds.php (hunk 3 above)
Comment (dguhl) [reply ds/ds.php:1349]: This is not intended to mark the callable as optional. What I forgot is to add the default value to an optional callable (null). The line you marked is the actual text of the documentation. https://www.php.net/manual/en/ds-map.filter.php

File: ds/ds.php [hunk 5]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+         * example, $map["1"] will attempt to access int(1), while $map->get("1") will correctly look up the string key.
+         *
+         * @param mixed $key The key to look up.
+         * @param mixed $default The optional default value, returned if the key could not be found.
+         *
+         * @return The value mapped to the given key, or the default value if provided and the key could not be found in the map.
+         *
+         * @throws OutOfBoundsException if the key could not be found and a default value was not provided.
+         */
+        public function get($key, $default)
```
Comment (isfedorov) [inline ds/ds.php:1389]: please add `null` as default value for `$default` as this parameter is optional

File: ds/ds.php [hunk 6]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+        public function filter(callable $callback): Map
+        {
+        }
+
+        /**
+         * Returns the first pair in the map
+         *
+         * @return Pair The first pair in the map.
+         *
+         * @throws UnderflowException if empty
```
Comment (isfedorov) [inline ds/ds.php:1365]: Could you please add link to doc?

File: ds/ds.php [hunk 7]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+         * <b>Caution:</b> Be careful when using array syntax. Scalar keys will be coerced to integers by the engine. For
+         * example, $map["1"] will attempt to access int(1), while $map->get("1") will correctly look up the string key.
+         *
+         * @param mixed $key The key to look up.
+         * @param mixed $default The optional default value, returned if the key could not be found.
+         *
+         * @return The value mapped to the given key, or the default value if provided and the key could not be found in the map.
+         *
+         * @throws OutOfBoundsException if the key could not be found and a default value was not provided.
+         */
```
Comment (isfedorov) [inline ds/ds.php:1388]: Could you please add link to doc?

File: ds/ds.php [hunk 8]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+         * A ∩ B = {x : x ∈ A ∧ x ∈ B}
+         *
+         * <p>
+         * <b>Note:</b> Values from the current instance will be kept.
+         *
+         * @param Map $map The other map, containing the keys to intersect with.
+         *
+         * @return Map The key intersection of the current instance and another map.
+         */
+        public function intersect(Map $map): Map
```
Comment (isfedorov) [inline ds/ds.php:1432]: Could you please add link to doc?

File: ds/ds.php (hunk 5 above)
Comment (dguhl) [reply ds/ds.php]: I did a bit more than that and re-checked the whole file for similar mistakes.

File: ds/ds.php [hunk 9]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+        }
+
+        /**
+         * Determines whether the map contains a given key
+         *
+         * @param mixed $key The key to look for.
+         *
+         * @return bool Returns TRUE if the key could found, FALSE otherwise.
+         */
+        public function hasKey($key): bool
```
Comment (isfedorov) [inline ds/ds.php:1402]: Could you please add link to doc?

File: ds/ds.php [hunk 10]
```diff
@@ -1272,21 +1326,126 @@ public function copy(): Collection
...
+        }
+
+        /**
+         * Determines whether the map contains a given value
+         *
+         * @param mixed $value The value to look for.
+         *
+         * @return bool Returns TRUE if the value could found, FALSE otherwise.
+         */
+        public function hasValue($value): bool
```
Comment (isfedorov) [inline ds/ds.php:1413]: Could you please add link to doc?

File: ds/ds.php [hunk 11]
```diff
@@ -1636,9 +1795,9 @@ public function sum()
          * Creates a new map using values from the current instance and another
          * map.
          *
-         * A ∪ B = {x: x ∈ A ∨ x ∈ B}
+         * A ∪ B = {<b>x:</b> x ∈ A ∨ x ∈ B}
```
Comment (isfedorov) [inline ds/ds.php:1814]: What is the purpose of selection of `x:`? Wasn't it made by mistake?

File: ds/ds.php [hunk 12]
```diff
@@ -2147,7 +2306,7 @@ public function sum()
          * Creates a new set that contains the values of the current instance as
          * well as the values of another set.
          *
-         * A ∪ B = {x: x ∈ A ∨ x ∈ B}
+         * A ∪ B = {<b>x:</b> x ∈ A ∨ x ∈ B}
```
Comment (isfedorov) [inline ds/ds.php:2325]: The same as above: What is the purpose of selection?

File: ds/ds.php (hunk 11 above)
Comment (dguhl) [reply ds/ds.php:1814]: Formatting was by mistake. I don't see how that could have happened, since the original docs don't have these marks. I removed it.

Comment (dguhl) [reply ds/ds.php]: By a regex (similar to `\w+:`), I included markings for "Notice:" and "Caution:" in the doc blocks and that must have got over "x:" too.

**Comment by dguhl** [comment]
My former PR has been squashed to keep the history meaningful and uncluttered. This one got merged despite all the minor corrections. :(

**Comment by isfedorov** [comment]
@dguhl I'm not sure that I got it. In scope of this PR all changes from https://github.com/JetBrains/phpstorm-stubs/pull/833/files have been merged into master branch of this repo. Could you please clarify what minor corrections haven't been merged?

**Comment by b1rdex** [comment]
> My former PR has been squashed to keep the history meaningful and uncluttered. This one got merged despite all the minor corrections. :(

@dguhl @isfedorov there is a 'squash and merge' merge scenario in GitHub interface:
image

**Comment by dguhl** [comment]
@isfedorov My contribution was merged completely, thank you. But I maybe wasn't clear: I meant my little correctional commits are now part of the main project's commit history, but they clutter the commit history. That's why I thought it's established practice to squash the PR in only one single commit before merging it, so the history stays clean?

**Comment by isfedorov** [comment]
@dguhl Oh, I see now. Yeah, commits could be squashed but in this PR I've decides that commits are quite good and every one describe quite well what was done in its scope and that's why I've decided not to merge them into one and rebase them as is.

**Comment by dguhl** [comment]
Alrighty. Thank you for the quick work on that.

**Comment by enumag** [comment]
@dguhl We found some more issues with ext-ds stubs in this package. See these issues in PHPStan repo for details:
https://github.com/phpstan/phpstan/issues/3539
https://github.com/phpstan/phpstan/issues/3513
https://github.com/phpstan/phpstan/issues/3512
Sadly I don't have time to fix it myself at the moment.

**Comment by dguhl** [comment]
Well then I take it from here. 😉

**Comment by dguhl** [comment]
I amended Data Structures for the third time in PR #856

**Pull request merged**